arangod/Aql/QuerySnippet.cpp

```diff
@@ -407,14 +441,36 @@ ResultT<std::unordered_map<ExecutionNode*, std::set<ShardID>>> QuerySnippet::pre
           // provide a correct translation from collection to shard
           // to be used in toVelocyPack methods of classes derived
           // from GraphNode
-          if (found->second == server || !localGraphNode->isDisjoint()) {
-            // TODO: Optimize this, we're looking into to many collections here
+          if (localGraphNode->isDisjoint()) {
+            if (found->second == server) {
+              myExp.emplace(shard);
+            }
+          } else {
             localGraphNode->addCollectionToShard(aqlCollection->name(), shard);
           }
 
           numShards++;
         }
+
+        if (myExp.size() > 0) {
+          myExpFinal.insert({aqlCollection->name(), std::move(myExp)});
+        }
+      }
+
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+      // additional verification checks for disjoint smart graphs
+      if (localGraphNode->isDisjoint()) {
+        size_t size = myExpFinal.begin()->second.size();
+        for (auto const& expDefinition : myExpFinal) {
```

Comment:
I think this loop contains wrong assertions.

The assertions should be the following:

// We need one expansion for every collection in the Graph
myExpFinal.size() == localGraphNode->collections().size()
for (exp : myExpFinal) {
  // Every of these expansions needs to have equal amount of shards.
  exp.second.size() == size
}

Reply:
resolved